### ISSUE TITLE
Windows: Crash on startup when SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\ does not exist in the registry

### ISSUE THREAD

**Issue by pineapplemachine** [Nexus-Mods/Nexus-Mod-Manager#384]
**Updated post:**

After much troubleshooting I was able to fix the problem where NMM crashed at startup. It turned out to be an issue with my .NET installation caused by repairing my Windows 7 installation. NMM runs correctly after installing Windows updates.

The error I got when NMM crashed was a very uninformative null reference exception, and having to determine myself where the exception may have taken place made troubleshooting a lot more difficult than it could have otherwise been. This error would have been essentially unsolvable for anyone who didn't have the knowledge needed to identify the location in the code base where the exception occurred.

I recommend adding error handling for the case where `SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full\` is missing from the registry. In this case, NMM should suggest that the user needs to either install .NET v4 or repair a faulty installation.

Someone with a Windows development environment set up could verify that I'm correct in thinking that the null reference exception was occurring at [L222 of Program.cs](https://github.com/Nexus-Mods/Nexus-Mod-Manager/blob/master/NexusClient/Program.cs#L222) when the aforementioned registry key is missing, causing the `ndpKey` variable to be null. After verifying that this was in fact the cause of the error I was experiencing, then a statement should be added above L222 throwing an exception with an explanatory message or else reporting the error in a more user-friendly way, e.g. in a popup alert.

**Original post:**

Nexus Mod Manager crashes with an error as soon as I attempt to open it, before any NMM window appears.

I recently installed the latest version of NMM to replace the older version I was using in the past (0.61.6). I kept the old version around when updating, and when I tried to run the older install I found that NMM 0.61.6 also crashes with the same error.

I was able to start up both programs without errors before I repaired my Windows 7 OS (using the "update" installation option) due to an unrelated issue with the Windows Installer service. I tried restarting my PC and reinstalling 0.65.2 after this and it did not fix the crashing problem.

``` text
Trace file has been created: TraceLog20180608151211.txt

Tracing an Unhandled Exception:
Exception: 
Message: 
	Object reference not set to an instance of an object.
Full Trace: 
	System.NullReferenceException: Object reference not set to an instance of an object.
   at Nexus.Client.Program.EnableTracing(IEnvironmentInfo p_eifEnvironmentInfo, Boolean p_booForceTrace)
   at Nexus.Client.Program.Main(String[] p_strArgs)
```

**Environment**
- NMM version: _0.65.2_
- Mod(s) used: N/A
- Operating System: Windows 7 64-bit

**Comment by pineapplemachine** [comment]
Forewarning: C# is not one of my stronger languages and I don't know anything about the registry. I might be chasing totally the wrong thing here.

I followed the stack trace to the EnableTracing function and found that the registry key `SOFTWARE\\Microsoft\\NET Framework Setup\\NDP\\v4\\Full\\`, referenced at [Program.cs#L220](https://github.com/Nexus-Mods/Nexus-Mod-Manager/blob/master/NexusClient/Program.cs#L220), does not exist on my machine. Underneath `NDP` I have only `v2.0.50726`, `v3.0`, and `v3.5`. I noticed in the documentation that the method call on this line [returns null in a failure case](https://msdn.microsoft.com/en-us/library/z9f66s0a\(v=vs.110\).aspx), (the documentation reads _If the specified subkey cannot be found, then null is returned_) yet line 222 accesses a member of the returned value without checking for null. I thought that this may be the cause of the NullReferenceException.

I attempted to install .NET v4.0 from the downloader [here](https://www.microsoft.com/en-us/download/confirmation.aspx?id=17851) thinking that perhaps the Windows repair had messed up a .NET v4 dependency. and installation fails with a ["fatal error"](https://i.imgur.com/Mj0tFPf.png). The error log ends with the failed installation of "Microsoft .NET Framework 4 Setup_4.0.30319\netfx_Core_x64.msi" but neither this log nor the MSI log gives a specific error. (Only a generic error code 1603.) However, when I run that installer independently it fails with an error saying ["A later version [...] is already installed."](https://i.imgur.com/VVcu0SU.png).

Next I tried a [v4.5 installer](https://www.microsoft.com/en-us/download/details.aspx?id=30653), which failed at an equivalent point with [an almost identical error message](https://i.imgur.com/jgkVXdf.png).

Now I'm going to try installing Windows updates and will update this post as troubleshooting progresses...

**Comment by squid-box** [comment]
NMM indeed requires .NET 4 , or maybe even 4.5 now (not sure). It does sound like the repair put your 4/4.5 in a weird state :(

**Comment by pineapplemachine** [comment]
Still going through the Windows update process to see if it affects the problem...

Yeah, I'm pretty sure this is an issue with my environment. It would be nice if the missing registry key (assuming that's in fact the issue) would produce a meaningful error message instead of just a very opaque null reference exception that I had to look through the source to guess at its meaning, though...

**Comment by pineapplemachine** [comment]
Good news! Installing Windows updates corrected the issue. (132 of them, goddamn. It took like two hours.) After updating Windows, the [aforementioned registry key _does_ exist](https://i.imgur.com/pUZQAhc.png), leading me to believe that I succeeded in identifying the specific cause of the crash.

So I'm going to update the original post and rename the issue to clarify that I was able to fix my problem, but that I believe the code should be updated to provide a more helpful error message for this failure case.
